# ------------- SLAM-SEQ ANALYSIS ------------
#
#   SLAM-Seq data analysis (SlamDunk) for paired-end reads
#
#   Author: Paulina Rosales Becerra (2022)

import os
import pandas as pd
import requests
import urllib3
from snakemake.utils import min_version
from snakemake.logging import logger

# Minimal version of snakemake
min_version('6.8.0')


# Load and read Snakemake config 
configfile: 'config/config_template.yaml'

# --------------------------- Read config file ---------------------------- #

RESULTS_DIR = Path(config['RESULTS_DIR'])
FIGURE_DIR = Path(config['FIGURE_DIR'])
QC_DIR = Path(config['QC_DIR'])

EXTERNAL_DATA_DIR = Path(config['EXTERNAL_DATA_DIR'])
FASTQ_DIR = Path(config['FASTQ_DIR'])

LOG_DIR = Path(config['LOG_DIR'])
TEMP_DIR = Path(config['TEMP_DIR'])
REPORT_DIR = Path(config['REPORT_DIR'])

SUBSET_COLUMN = config['SUBSET_COLUMN']


# --------------------------- Read sample manifest ---------------------------- #
SAMPLE_MANIFEST = pd.read_table(Path(config['SAMPLE_MANIFEST']))

# Specify the sample info combinations to use as label per sample
SAMPLES = SAMPLE_MANIFEST.set_index(['Sample_type',
                                     'Treatment',
                                     'Bio_rep'])


# --------------------------- Dictitonaries ---------------------------- #

# GENCODE releases dictionary (release -> genome name # UCSC genome equivalent)
GENCODE_VERSIONS = {
    'M25': 'GRCm38', # mm10
    'M32': 'GRCm39', # mm39
    '19': 'GRCh37', # h19
    '38': 'GRCh38', # h38 
    # 'M1': 'NCBI37' # mm9 uses different handle
}

# Output files dictionary
TARGETS = {'raw_seq': [], 
            'trimmed': [],
            'slam_map': [],
            'slam_filter': [],
            'slam_snp': [],
            'slam_count': [],
            'gene_counts': []
            }

# as array for different genomes?
GENCODE_RELEASE = pd.unique(SAMPLES['Target_genome'])[0]
GENOME = GENCODE_VERSIONS[GENCODE_RELEASE]

for sample_type, treatment, bio_rep in SAMPLES.index:
    # Select samples with specified value on a given column in the sample manifest (USE CONFIG FILE TO SELECT)
    if SUBSET_COLUMN:
        if SAMPLES.loc[(sample_type, treatment, bio_rep), SUBSET_COLUMN] == config['SUBSET_VALUE']: 
            TARGETS['raw_seq'].append(str(FASTQ_DIR / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_R1.fq.gz'))
            TARGETS['trimmed'].append(str(RESULTS_DIR / 'fastq' / 'trimmed' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_R1.fq.gz_trimming_report.txt'))
            TARGETS['slam_map'].append(str(RESULTS_DIR / 'bam_files' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.bam'))
            TARGETS['slam_filter'].append(str(RESULTS_DIR / 'bam_files' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_filtered.bam.bai'))
            TARGETS['slam_snp'].append(str(RESULTS_DIR / 'variant_call' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_filtered_snp.vcf'))
            TARGETS['slam_count'].append(str(RESULTS_DIR / 'counts' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_filtered_tcount.tsv'))
            TARGETS['gene_counts'].append(str(RESULTS_DIR / 'counts' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.genecount.tsv'))
    else:
            TARGETS['raw_seq'].append(str(FASTQ_DIR / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_R2.fq.gz'))
            TARGETS['trimmed'].append(str(RESULTS_DIR / 'fastq' / 'trimmed' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_R1.fq.gz_trimming_report.txt'))
            TARGETS['slam_map'].append(str(RESULTS_DIR / 'bam_files' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.bam'))
            TARGETS['slam_filter'].append(str(RESULTS_DIR / 'bam_files' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_filtered.bam.bai'))
            TARGETS['slam_snp'].append(str(RESULTS_DIR / 'variant_call' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_filtered_snp.vcf'))
            TARGETS['slam_count'].append(str(RESULTS_DIR / 'counts' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_filtered_tcount.tsv'))
            TARGETS['gene_counts'].append(str(RESULTS_DIR / 'counts' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.genecount.tsv'))
            # TARGETS['alley_rates'].append(str(RESULTS_DIR / 'conversion_rates' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_overallrates.csv'))
            # TARGETS['alley_utrrates'].append(str(RESULTS_DIR / 'conversion_rates' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_mutationrates_utr.csv'))
            # TARGETS['dif_expr'].append(str(RESULTS_DIR / 'dif_expression' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_mutationrates_utr.csv'))
            # TARGETS['figures'].append(str(RESULTS_DIR / 'conversion_rates' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_mutationrates_utr.csv'))
            
    

# --------------------------- Import rules ---------------------------- #

# --------- Raw procesing 
include: 'rules/external_data/get_refseq.smk'
include: 'rules/mapping/trim_galore.smk'
include: 'rules/mapping/mapping.smk'
include: 'rules/counts/read_counts.smk'
include: 'rules/external_data/annotation_tables.smk'
include: 'rules/counts/collapse_counts.smk'

# --------- Downstream analysis
# include: 'rules/plots.smk'
# include: 'rules/conversion_rates.smk'
# include: 'rules/downstream/DE_analysis.smk'

# --------------------------- Rules ---------------------------- #


# rule multiqc_report:
#     output:
#         base = directory(str(REPORT_DIR / 'multiqc'))
#     params:
#         title = 'MultiQC report: SLAM-Seq data'
#     conda:
#         'envs/quality_control/multiqc.yaml'
#     log:
#         str(LOG_DIR / 'multiqc' / 'multiqc.log')
#     shell:
#         """
#         multiqc -o {output.base} -n multiqc_report.html\
#         -m fastqc -m cutadapt -m slamdunk -f\
#         --title {params.title}\
#         {REPORT_DIR} {RESULTS_DIR} {LOG_DIR} &> {log}
#         """
#     shell:
#         """
#         multiqc --force --interactive \
#         -o {output.base} -n multiqc_report.html \
#         {REPORT_DIR} {RESULTS_DIR} {LOG_DIR} &> {log}
#         """


rule all:
    input:
        expand('resources/external/gencode_{release}/{genome}.genome.fa.gz', release=GENCODE_RELEASE, genome=GENOME),
        TARGETS['raw_seq'],
        TARGETS['trimmed'],
        TARGETS['slam_map'],
        TARGETS['slam_filter'],
        TARGETS['slam_snp'],
        TARGETS['slam_count'],
        TARGETS['gene_counts']
    # default_target: True
